File: 2020/csharp/Aoc2020.Tests/Day08Tests.cs
using System.Linq;
using System.Collections.Generic;
using Xunit;

namespace Aoc2020.Tests
{
    public class Day08Tests
    {
        [Fact]
        public void InfiniteLoopDetectionWorks()
        {
            var code = Code.Parse(Utils.ReadInputLines("08-test"));
            code.Run();
            Assert.Equal(5, code.Accumulator);
        }

        [Fact]
        public void MutateCodeWorks()
        {
            Assert.Equal(8, Day08.MutateCode(Utils.ReadInputLines("08-test")).Accumulator);
        }

        [Fact]
        public void Stars()
        {
            var run = new Day08();
            Assert.Equal("1753", run.Result1());
            Assert.Equal("733", run.Result2());
        }
    }
}

File: 2022/fsharp/Aoc2022.Tests/Day08.Tests.fs
module Day08.Tests

open Xunit
open Utils
open Day08

let input =
    """
    30373
    25512
    65332
    33549
    35390
    """
    |> String.trim
    |> String.split '\n'
    |> Array.map String.trim
    |> Grid.parse

[<Fact>]
let ``Day08 countVisibleTrees works`` () =
    Assert.Equal(21, input |> countVisibleTrees)

[<Fact>]
let ``Day08 maxScenicScore works`` () =
    Assert.Equal(8, input |> maxScenicScore)

[<Fact>]
let ``Day08 Stars`` () =
    try
        Assert.Equal("1789", job1 ())
        Assert.Equal("314820", job2 ())
    with :? System.NotImplementedException ->
        ()
